internal/controller/apiextensions/composite/watch/watch.go

```diff
@@ -0,0 +1,143 @@
+
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	"github.com/crossplane/crossplane-runtime/pkg/resource/unstructured/composite"
+)
+
+// A ControllerEngine can stop watches for a controller, and remove informers.
+type ControllerEngine interface {
+	StopWatches(ctx context.Context, name string, keep ...schema.GroupVersionKind) error
+	RemoveUnwatchedInformers(ctx context.Context) error
+}
+
+// A GarbageCollector garbage collects watches for a composite resource
+// controller.
+type GarbageCollector struct {
+	controllerName string
+	xrGVK          schema.GroupVersionKind
+
+	c  client.Reader
+	ce ControllerEngine
+
+	alwaysKeep []schema.GroupVersionKind
+
+	log logging.Logger
+}
+
+// A GarbageCollectorOption configures a GarbageCollector.
+type GarbageCollectorOption func(gc *GarbageCollector)
+
+// WithLogger configures how a GarbageCollector should log.
+func WithLogger(l logging.Logger) GarbageCollectorOption {
+	return func(gc *GarbageCollector) {
+		gc.log = l
+	}
+}
+
+// AlwaysKeep specifies a set of GVKs that the garbage collector should always
+// keep - i.e. should never garbage collect. Typically this will be the XR's
+// GVK, and that of a CompositionRevision.
+func AlwaysKeep(gvks ...schema.GroupVersionKind) GarbageCollectorOption {
+	return func(gc *GarbageCollector) {
+		gc.alwaysKeep = gvks
+	}
+}
+
+// NewGarbageCollector creates a new watch garbage collector for a controller.
+func NewGarbageCollector(name string, of resource.CompositeKind, c client.Reader, ce ControllerEngine, o ...GarbageCollectorOption) *GarbageCollector {
+	gc := &GarbageCollector{
+		controllerName: name,
+		xrGVK:          schema.GroupVersionKind(of),
+		c:              c,
+		ce:             ce,
+		log:            logging.NewNopLogger(),
+	}
+	for _, fn := range o {
+		fn(gc)
+	}
+	return gc
+}
+
+// GarbageCollectWatches runs garbage collection at the specified interval,
+// until the supplied context is cancelled. It stops any watches for resource
+// types that are no longer composed by any composite resource (XR). It also
+// removed any informers that no loner power any watches.
+func (gc *GarbageCollector) GarbageCollectWatches(ctx context.Context, interval time.Duration) {
+	t := time.NewTicker(interval)
+	defer t.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			gc.log.Debug("Stopping composite controller watch garbage collector", "error", ctx.Err())
+			return
+		case <-t.C:
+			if err := gc.GarbageCollectWatchesNow(ctx); err != nil {
+				gc.log.Info("Cannot garbage collect composite controller watches", "error", err)
+			}
+		}
+	}
+}
+
+// GarbageCollectWatchesNow stops any watches for resource types that are no
+// longer composed by any composite resource (XR). It also removed any informers
+// that no longer power any watches.
+func (gc *GarbageCollector) GarbageCollectWatchesNow(ctx context.Context) error {
```

Comment:
is this supposed to be go routine safe? If not, make it private?

Reply:
It is - added a comment to clarify.